File: 纪中刘远哲/2015/4.11/1237.cpp
#include <cstdio>
#include <algorithm>
using namespace std;

const int N = 2001;
int n , m , ans;
char st[N][N];
//int sum[N][N];
int up[N][N];

int main(){
	freopen("3.in","r",stdin);
	scanf("%d %d" , &n , &m);
	for(int i = 1; i <= n; i ++) scanf("%s" , &st[i]);
	for(int i = 1; i <= n; i ++)
		for(int j = 1; j <= m; j ++){
			if (st[i][j - 1] == 'X') up[i][j] = 0;
			else up[i][j] = up[i - 1][j] + 1;
		}
	
	int i = n;
	ans = 0;
	while(i >= 1){
		for(int j = 1; j <= m; j ++){
			int low = up[i][j];
			for(int k  = j; k <= m; k ++){
				low = min(low , up[i][k]);
				if(low == 0) break;
				ans = max(ans , (k - j + 1 + low) * 2);
			}
		}
		i --;
	}
	
	printf("%d" , ans - 1);
	
	return 0;
}

File: OI/纪中刘远哲/2015/4.11/1237.pas
uses	math;
const	maxn=2000;maxm=2000;
var	i,j,k,m,n,ans,low:longint;
	up:array[0..maxn,0..maxm] of longint;
	s:ansistring;		//change
begin
	assign(input,'3.in');reset(input);
	//assign(output,'.out');rewrite(output);

	readln(n,m);
	for i:=1 to n do begin
		readln(s);
		for j:=1 to m do begin
			if s[j]='X' then up[i,j]:=0 else 
				up[i,j]:=up[i-1,j]+1;
		end;
	end;

	ans:=0;
	for i:=n downto 1 do 
		for j:=1 to m do begin
			low:=up[i,j];
			for k:=j to m do begin
				low:=min(low,up[i,k]);
				if low=0 then break;
				ans:=max(ans,(k-j+1+low)*2);
			end;
		end;
	writeln(ans-1);
end.
